service/datastore/client-rest/src/main/java/org/eclipse/kapua/service/datastore/client/rest/RestDatastoreClient.java

```diff
@@ -0,0 +1,641 @@
+                }
+            } else if (esClientProvider == null) {
+                // something wrong happened in the client lifecycle... try to cleanup and raise exception
+                cleanupClient(true);
+            }
+        }
+    }
+
+    private void cleanupClient(boolean raiseException) throws ClientUnavailableException {
+        Throwable cause = null;
+        try {
+            esClientProvider.getClient().close();
+            esClientProvider = null;
+            instance = null;
+        } catch (Throwable e) {
+            cause = e;
+            logger.error(CLIENT_CLEANUP_ERROR_MSG, e);
+        }
+        if (cause != null) {
+            throw new ClientUnavailableException(CLIENT_CLEANUP_ERROR_MSG, cause);
+        } else if (raiseException) {
+            throw new ClientUnavailableException(CLIENT_CLEANUP_ERROR_MSG);
+        }
+    }
+
+    @Override
+    public InsertResponse insert(InsertRequest insertRequest) throws ClientException {
+        checkClient();
+        try {
+            Map<String, Object> storableMap = modelContext.marshal(insertRequest.getStorable());
+            logger.debug("Insert - converted object: '{}'", storableMap.toString());
+            String json = MAPPER.writeValueAsString(storableMap);
+            HttpEntity entity = new NStringEntity(json, ContentType.APPLICATION_JSON);
+            Response insertResponse = esClientProvider.getClient().performRequest(
+                    POST_ACTION,
+                    getTypePath(insertRequest.getTypeDescriptor()),
+                    Collections.<String, String>emptyMap(), 
+                    entity);
+            if (isRequestSuccessful(insertResponse)) {
+                JsonNode responseNode = MAPPER.readTree(EntityUtils.toString(insertResponse.getEntity()));
+                String id = ((TextNode) responseNode.get(KEY_DOC_ID)).asText();
+                String index = ((TextNode) responseNode.get(KEY_DOC_INDEX)).asText();
+                String type = ((TextNode) responseNode.get(KEY_DOC_TYPE)).asText();
+                return new InsertResponse(id, new TypeDescriptor(index, type));
+            }
+            else {
+                throw new ClientException(ClientErrorCodes.ACTION_ERROR, insertResponse.getStatusLine().getReasonPhrase());
+            }
+        } catch (IOException e) {
+            throw new ClientException(ClientErrorCodes.ACTION_ERROR, e, e.getLocalizedMessage());
+        }
+    }
+
+    @Override
+    public UpdateResponse upsert(UpdateRequest updateRequest) throws ClientException {
+        checkClient();
+        try {
+            Map<String, Object> storableMap = modelContext.marshal(updateRequest.getStorable());
+            Map<String, Object> updateRequestMap = new HashMap<>();
+            updateRequestMap.put(KEY_DOC, storableMap);
+            updateRequestMap.put(KEY_DOC_AS_UPSERT, true);
+            logger.debug("Upsert - converted object: '{}'", updateRequestMap.toString());
```

Comment:
Remove `.toString()` is already called `.toString()` for each parameter passed as argument.